### ISSUE TITLE
Layer in TimeDistributed fails to unserialize if also used elsewhere

### ISSUE THREAD

**Issue by FremyCompany** [tensorflow/tfjs#1538]
#### TensorFlow.js version
`1.0.2`

#### Describe the problem or feature request
When you create a model in which a layer is used in more than one TimeDistributed layer, serialization will duplicate the layer as if they were two different layers, and then unserialization will fail because only one weight will be serialized for the "two" layers.

#### Code to reproduce the bug / link to feature request
```js

    /** @type {tf.LayersModel} */
    let model; {

        let w1_inp = tf.input({ shape: [1], dtype: 'int32' });
        let w2_inp = tf.input({ shape: [1], dtype: 'int32' });
        let w_inp = tf.input({ shape: [NUMBER_OF_ALTERNATIVES, 1], dtype: 'int32' });

        let embedding_words = tf.layers.embedding({ 
            inputDim: vocab_map.size, outputDim: EMBEDDING_SIZE, 
            embeddingsRegularizer: null,
            activityRegularizer: null,
            embeddingsConstraint: tf.constraints.minMaxNorm({ axis:1, minValue:0.25, maxValue:4.0 }),
            //embeddingsInitializer: 'zeros',
            inputShape: [1], inputLength: 1, 
        })

        let w = tf.layers.timeDistributed({ name: 'words_embeddings', layer: embedding_words }).apply(w_inp);
        let w1 = embedding_words.apply(w1_inp);
        let w2 = embedding_words.apply(w2_inp);

        // ...

        model = tf.model({ inputs: [w1_inp,w2_inp,w_inp], outputs: [softmax] });
    };

    // ...

    model.save(`file://${model_save_dir}`);

```

**Comment by FremyCompany** [comment]
The only way to fix this I can imagine would be to allow the `layer` property of a TimeDistributed layer to take a string instead of an object as an input, and when this happens, lookup for an already unserialized layer with that name and use it as is, and not create a new layer.

**Comment by FremyCompany** [comment]
(It does seem that giving a name to the layer works, but I suspect that there are now two layers with the same name in the model and that is bound to cause issues somewhere at a later point, plus that means it's probably not possible to train the model anymore since the two copies would diverge)

**Comment by FremyCompany** [comment]
UPDATE: for what this is worth, I got things working on my end with some code hacking; here is how I got things working:

1. **PASS THE LIST OF LAYERS:** In `container.ts` append a `getCreatedLayers` function to `customObjects` which returns `createdLayers` which stores all the layers that have already been created

2. **RETREIVE FROM THE LIST OF LAYERS:** In `wrapper.ts` in `Wrapper.fromConfig` get `layerConfigName` then attempt to get the layer from `customObjects.getCreatedLayers()[layerConfigName]` (and only call `deserialize` if that doesn't work)

3. **MAYBE UPDATE SERIALIZATION CODE:** I also renamed the variable name in `model.json` to match what the model really expected, so there is probably some change needed somewhere in the serialization code.

I could do a PR following those lines if the maintainers thinks this is an acceptable approach.

**Comment by caisq** [comment]
@FremyCompany Thanks for filing the issue. To make sure I understand your original intent, why do you need to apply `timeDistributed` on `embedding`? `embedding` supports look up along the time axis on its own.

**Comment by FremyCompany** [comment]
@caisq Yes, but I need to specify `inputLength` to a fixed number, so that I can flatten the things later. But I also need to apply the embedding on sequences of differing sizes so all in all, I can't use the built-in functionality. I could have two embeddings of different inputLength yet sharing the same weight, but that seemed difficult to do; choosing a sequence of size one then applying TimeDistributed myself seemed to fit the bill better.

I would suspect the issue would reproduce for any kind of layer not just an embedding layer though. Any layer that you use both inside a TimeDistributed and outside of it would probably face the same problem.

**Comment by rthadur** [comment]
Automatically closing due to lack of recent activity. Please update the issue when new information becomes available, and we will reopen the issue. Thanks!
